_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: post
+title: "Model-View-ViewModel Pattern in Android"
+description: Our choice for the upday app - the Model-View-ViewModel pattern.
+modified:
+categories: blog
+author: florina_muntenescu
+excerpt: Our choice for the upday app - the Model-View-ViewModel pattern. Find out what it is, how we applied it in Android and why we consider it perfect for us.
+tags: [Android, Architecture, MVVM]
+date: 2016-09-04T15:39:55-04:00
+---
+After four different designs in the first six months of the development of the upday app, we learned one important lesson: we need an architecture pattern that allows fast reaction to design changes! The solution we chose in the end was Model-View-ViewModel. Discover with me what MVVM is; how we are applying it at upday and what makes it so perfect for us.
+
+## The Model-View-ViewModel Pattern
+
+The main players in the MVVM pattern are:
+
+* The *View* - that informs the ViewModel about the user's actions
+* The *ViewModel* - exposes streams of data relevant to the View
+* The *DataModel* - abstracts the data source. The ViewModel works with the DataModel to get and save the data.  
+
+At a first glance, MVVM seems very similar to the Model-View-Presenter pattern, because both of them do a great job in abstracting the view's state and behavior. The Presentation Model abstracts a view independent from a specific user-interface platform whereas the MVVM pattern was created by Microsoft to simplify the **event driven** programming of user interfaces.
+
+If, in the MVP pattern, the Presenter was telling the View directly what to display, in MVVM, **ViewModel exposes streams of events** to which the Views can bind to. Also, the Views notify the ViewModel about different actions. Thus, the MVVM pattern supports two-way data binding between the View and ViewModel and there is a many-to-one relationship between View and ViewModel. View has a reference to ViewModel but **ViewModel has no information about the View**.
+
+<center>
+<picture>
+	<a href="/images/blog/model_view_viewmodel/mvvm.png"><img src="/images/blog/model_view_viewmodel/mvvm.png" alt="Model-View-ViewModel Architecture"></a>
+	<figcaption>Model-View-ViewModel class structure</figcaption>
+</picture>
+</center>
+
+## Model-View-ViewModel at upday
+
+A quick look at the Android posts on the upday blog will instantly reveal what our favorite library is: RxJava. So it’s no wonder that RxJava is the backbone of upday's code. The event driven part required by MVVM is done using RxJava’s `Observable`s. Here's how we apply MVVM at upday, with the help of RxJava:
```

Comment:
By saying:
> Here's how we apply MVVM at upday

It's making more of a reference the entire company, rather than the Android app. Was that the intent?